src/java.base/share/classes/java/lang/reflect/code/op/ExtendedOp.java

```diff
@@ -3155,37 +3155,117 @@ static Block.Builder lowerTypePattern(Block.Builder endNoMatchBlock, Block.Build
+
                 bindings.add(target);
 
                 return currentBlock;
             }
 
+            private static boolean isUnboxingFollowedByPrimitiveWidening(TypeElement s, TypeElement t) {
+                return s instanceof ClassType ct && ct.unbox().isPresent() && t instanceof PrimitiveType pt &&
+                        !pt.box().orElseThrow().equals(s);
+            }
+
+            private static boolean isNarrowingRefConvFollowedByUnboxing(TypeElement s, TypeElement t) {
+                return s instanceof ClassType ct && ct.unbox().isEmpty() && t instanceof PrimitiveType;
+            }
+
+            private static boolean isUnboxing(TypeElement s, TypeElement t) {
+                return s instanceof ClassType && t instanceof PrimitiveType pt && pt.box().orElseThrow().equals(s);
+            }
+
+            private static boolean isIdentityPrimitiveConv(TypeElement s, TypeElement t) {
+                return s instanceof PrimitiveType && t instanceof PrimitiveType && Objects.equals(s, t);
+            }
+
+            private static boolean isWideningPrimitiveConvAndNeedsCheck(TypeElement s, TypeElement t) {
+                return (INT.equals(s) && FLOAT.equals(t)) ||
+                        (LONG.equals(s) && FLOAT.equals(t)) ||
+                        (LONG.equals(s) && DOUBLE.equals(t));
+            }
+            private static MethodRef convMethodRef(TypeElement s, TypeElement t) {
+                if (BYTE.equals(s) || SHORT.equals(s) || CHAR.equals(s)) {
+                    s = INT;
+                }
+                String sn = capitalize(s.toString());
+                String tn = capitalize(t.toString());
+                String mn = "is%sTo%sExact".formatted(sn, tn);
+                JavaType exactConversionSupport = JavaType.type(ClassDesc.of("java.lang.runtime.ExactConversionsSupport"));
+                return MethodRef.method(exactConversionSupport, mn, BOOLEAN, s);
+            }
+
+            private static boolean isWideningPrimitiveConv(TypeElement s, TypeElement t) {
+                if (!(s instanceof PrimitiveType sp) || !(t instanceof PrimitiveType tp)) {
+                    return false;
+                }
+                List<PrimitiveType> l = List.of(BYTE, SHORT, CHAR, INT, LONG, FLOAT, DOUBLE);
+                if (BYTE.equals(s) && CHAR.equals(t)) {
+                    return false;
+                } else if (SHORT.equals(s) && CHAR.equals(t)) {
+                    return false;
+                }
+                int si = l.indexOf(s);
+                int ti = l.indexOf(t);
+                return si < ti;
+            }
+
+            private static boolean isNarrowingPrimitiveConv(TypeElement s, TypeElement t) { // s -> t
+                if (!(s instanceof PrimitiveType sp) || !(t instanceof PrimitiveType tp)) {
+                    return false;
+                }
+                List<PrimitiveType> l = List.of(BYTE, SHORT, CHAR, INT, LONG, FLOAT, DOUBLE);
```

Comment:
We should avoid the allocation here on each call. Also i think you really need a map of bit sizes, then the special behaviour for short and char becomes clearer.